docs/rfc/rfc-027-p2p-message-bandwidth-report.md

```diff
@@ -0,0 +1,137 @@
+
+Beyond ensuring that transactions are rebroadcast to peers less frequently, there is not a simple scheme to communicate fewer transactions to peers. Peers cannot communicate what transactions they need since they do not know which transactions exist on the network.
+
+#### Vote Transmission
+
+Tendermint votes, both prevotes and precommits, are central to Tendermint consensus and are gossiped by all nodes to all peers during each consensus round. Data collected from the Blockpane node running on Osmosis indicates that the validator sounds about 9 gigabytes of `Vote` messages were sent during the nearly 3 hour period of observation.
+
+The Tendermint consensus reactor starts a new [gossipVotesRoutine][gossip-votes-routine] for each peer that it connects to. The reactor sends all votes to all peers unless it knows that the peer already has the vote or the reactor learns that the peer is in a different round and that thus the vote no longer applies. Tendermint learns that a peer has a vote in one of XX ways:
+
+1. Tendermint sent the peer the vote.
+2. Tendermint received the vote from the peer.
+3. The peer [sent a `HasVote` message][apply-has-vote]. This message is broadcast to all peers [each time validator receives a vote][publish-event-vote].
+4. The peer [sent a `VoteSetBits` message][apply-vote-set-bits]. This message is [sent as a response to a peer that sends a `VoteSetMaj23`][vote-set-bits-send].
+
+Given that Tendermint informs all peers of _each_ vote message it receives, all nodes should be well informed of which votes their peers have. Given that the vote messages were the third largest consumer of bandwidth in the observation on Osmosis, it's possible that this system is not currently working correctly. Further analysis should examine where votes may be being retransmitted.
+
+### Suggested Improvements to Lower Message Transmission Bandwidth
+
+#### Gossip Known BlockPart Data
+
+The `BlockPart` messages, by far, account for the majority of the data sent to each peer. At the moment, peers do not inform the node of which block parts they already have. This means that each block part is _very likely_ to be transmitted many times to each node. This frivolous consumption is even worse in networks large blocks. 
+
+The very simple solution to this issue is to copy the technique used in consensus for informing peers when the node receives a vote. The consensus reactor can be augmented with a `HasBlockPart` message that is broadcast to each peer every time the node receives a block part. By informing each peer every time the node receives a block part, we can drastically reduce the amount of duplicate data sent to each node. There would be no algorithmic way of enforcing that a peer accurately reports its block parts, so providing this message would be a somewhat altruistic action on the part of the node.
```

Comment:
This definitely makes sense :+1: It might be nice to measure how often we receive duplicate block parts (as well as duplicate transactions or votes) to understand how much of an impact this is

Reply:
Yes, this is an old proposal: https://github.com/tendermint/tendermint/issues/627